### DISCUSSION TITLE
zpool import -T hangs

### DISCUSSION THREAD

**Discussion by Wondertan** [openzfs/zfs#12758] [Q&A] [answered]
Type | Version/Name
 --- | ---
Distribution Name	| Arch
Kernel Version	| 5.14.15-arch1-1
Architecture	| amd64
OpenZFS Version	| zfs-kmod-2.1.1-1

Hello everyone. I have two FSs mounted on:
* `/dir`
* `/dir/subdir`

I had to migrate `/dir` to another FS and execute `mv /dir <new_path>`. Unfortunately, that also touched `/dir/subdir` with terabytes of data. Once I realized that the `/subdir` is getting moved as well I immediately interrupted the `mv`. However, since `/subdir` contains more subdirectories, critical files there were deleted during the move due to the nature of `mv`. 

To recover deleted files. I looked into `zpool history` and found valid txg to execute `zpool import -T`. Unfortunately, it hangs like for an hour already with no symptoms of life.

Pls note:
* I don't have snapshots
* It is impossible to get data from the place where it was moved.
* _Nothing_ was written to the pool under `/dir/subdir` after the move, so I hope it is possible to recover deleted data.
* The moved data is absolutely critical

Any ideas on why it may hang and maybe any tips on resolving the issue?

**Comment by rincebrain** [comment, marked as answer]
Using -T/-X/... by default walks every single referenced metadata+data block before importing. This is rather time consuming, and I don't think it uses anything like the sequential scrub logic to reorder things, so it's going to be similar to the runtime of the pre-sequential scrub code.

You can adjust this behavior with the tunables `spa_load_verify_data` (which can turn off reading all data blocks) and `spa_load_verify_metadata` (which will turn off reading all metadata or data first) - of course, you can run into inconsistent things on-disk if you don't verify them before importing, so it's a tradeoff.

I would pretty strongly recommend using -o readonly=on when playing with -T/-F/-X, though if you've already done it without doing that, you may have already done some damage...well, we'll see.

As the man page says, "WARNING: This option can be extremely hazardous to the health of your pool and should only be used as a last resort.", so be aware, it can go awry.

You said, though that there's "no symptoms of life" - does that mean no output (that's expected), or no IO on the pool at all (not so much). Is there anything in `dmesg` that might suggest it found something upsetting on trying to import and died?
